### PR TITLE
introduce global lock directory

### PR DESCRIPTION
**Pull request by xu-cheng**
Since #292, HOMEBREW_CACHE was moved to a per-user directory. This makes
it unsuitable to store global lock files on multiple users environment.

Therefore, introducing a global lock directory `/Library/Lock.d` to
store lock files from formula lockers as well as `brew update`.

### CODE REVIEW

File: Library/Homebrew/utils/lock.sh [hunk 1]
```diff
@@ -3,7 +3,7 @@
 # Noted due to the fixed FD, a shell process can only create one lock.
 lock() {
   local name="$1"
-  local lock_file="/tmp/homebrew${HOMEBREW_REPOSITORY//\//-}-${name}.lock"
+  local lock_file="$HOMEBREW_LIBRARY/Lock.d/$name"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/lock.sh]: Maybe just `$HOMEBREW_LIBRARY/Locks` instead? The `.d` looks weird to me.

**Comment by MikeMcQuaid** [comment]
Tiny comment otherwise 👍

**Comment by xu-cheng** [comment]
Updated.

**Comment by ilovezfs** [comment]
Am I misreading it, or is the default actually `/usr/local/Library/Locks`. I assume `/Library/Lock.d` is just a typo in the commit message?

**Comment by xu-cheng** [comment]
Oops, forget to update commit message.

**Comment by apjanke** [comment]
Minor comment: I'm finding the code a bit hard to follow, mostly as a result of the existing initialization logic.

There's a one-to-many relationship between a `HOMEBREW_REPOSITORY` and `HOMEBREW_PREFIX`es. (AFAIK.) It's not very clear from the code whether this lock dir is per-repo or per-prefix. I think this is because existing the `HOMEBREW_LIBRARY_PATH` is initialized from a source-relative path rather than one of `HOMEBREW_REPOSITORY` or `HOMEBREW_PREFIX`.

```
unless defined? HOMEBREW_LIBRARY_PATH
  HOMEBREW_LIBRARY_PATH = Pathname.new(__FILE__).realpath.parent
end
```

My reading of this PR is that the locks are per-repo, not per-prefix, and I think that's how it should be. But it took a bit of reading and thinking to figure that out. Maybe we could rewrite `HOMEBREW_LIBRARY_PATH` to be in terms of `HOMEBREW_REPOSITORY` or `HOMEBREW_PREFIX` instead of  being relative to `config.rb`? Would fit well in this PR, imho.

**Comment by xu-cheng** [comment]
@apjanke I think you may mistake the variable with `HOMEBREW_LIBRARY` and `HOMEBREW_LIBRARY_PATH`
